presto-thrift/src/main/java/com/facebook/presto/thrift/pagesources/ThriftIndexPageSource.java

```diff
@@ -0,0 +1,219 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.thrift.pagesources;
+
+import com.facebook.presto.spi.ColumnHandle;
+import com.facebook.presto.thrift.ThriftConfig;
+import com.facebook.presto.thrift.clientproviders.PrestoClientProvider;
+import com.facebook.presto.thrift.interfaces.client.ThriftPrestoClient;
+import com.facebook.presto.thrift.interfaces.client.ThriftRowsBatch;
+import com.facebook.presto.thrift.interfaces.client.ThriftSplit;
+import com.facebook.presto.thrift.interfaces.client.ThriftSplitBatch;
+import com.facebook.presto.thrift.interfaces.client.ThriftSplitsOrRows;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.facebook.presto.thrift.interfaces.client.ThriftHostAddress.toHostAddressList;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static com.google.common.util.concurrent.Futures.transform;
+import static com.google.common.util.concurrent.Futures.transformAsync;
+import static java.util.Objects.requireNonNull;
+
+public class ThriftIndexPageSource
+        extends ThriftAbstractPageSource
+{
+    private final PrestoClientProvider clientProvider;
+    private final ThriftRowsBatch keys;
+    private final byte[] indexId;
+    private final int maxSplitsPerBatch;
+
+    // modified from other threads
+    private AtomicReference<ThriftPrestoClient> client;
+    private AtomicReference<State> state;
+
+    public ThriftIndexPageSource(
+            PrestoClientProvider clientProvider,
+            byte[] indexId,
+            ThriftRowsBatch keys,
+            List<ColumnHandle> columns,
+            ThriftConfig config)
+    {
+        super(columns, config);
+        this.clientProvider = requireNonNull(clientProvider, "clientProvider is null");
+        this.keys = requireNonNull(keys, "keys is null");
+        this.indexId = requireNonNull(indexId, "indexId is null");
+        this.maxSplitsPerBatch = requireNonNull(config, "config is null").getMaxIndexSplitsPerBatch();
+        this.client = new AtomicReference<>(clientProvider.connectToAnyHost());
+        this.state = new AtomicReference<>(new InitialState());
+    }
+
+    @Override
+    public ListenableFuture<ThriftRowsBatch> sendRequestForData(byte[] nextToken, long maxBytes)
+    {
+        return state.get().sendRequestForData(nextToken, maxBytes);
+    }
+
+    @Override
+    public boolean canGetMoreData(byte[] nextToken)
+    {
+        return state.get().canGetMoreData(nextToken);
+    }
+
+    @Override
+    public void closeInternal()
+    {
+        client.get().close();
+    }
+
+    private interface State
+    {
+        ListenableFuture<ThriftRowsBatch> sendRequestForData(byte[] nextToken, long maxBytes);
+
+        boolean canGetMoreData(byte[] nextToken);
```

Comment:
I'd like to keep the current name, because it says that it's possible to request data, but not necessarily that the data is there.